File: example/Exam01Push/app/src/main/java/com/gbox3d/exam01push/MainActivity.java
package com.gbox3d.exam01push;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NotificationCompat;

import android.app.ActivityManager;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.InstanceIdResult;

public class MainActivity extends AppCompatActivity {

    public final static String TAG="Exam01";


    private void findToken()
    {
        FirebaseInstanceId.getInstance().getInstanceId()
                .addOnCompleteListener(new OnCompleteListener<InstanceIdResult>() {
                    @Override
                    public void onComplete(@NonNull Task<InstanceIdResult> task) {
                        if (!task.isSuccessful()) {
                            Log.w(TAG, "getInstanceId failed", task.getException());
                            return;
                        }

                        // Get new Instance ID token
                        String token = task.getResult().getToken();
                        Log.d(TAG,token);

                        // Log and toast
                        //String msg = getString(R.string.msg_token_fmt, token);
                        //Log.d(TAG, msg);
                        //Toast.makeText(MainActivity.this, msg, Toast.LENGTH_SHORT).show();
                    }
                });

    }

    private void createNotification() {

        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, "default");

        builder.setSmallIcon(R.mipmap.ic_launcher);
        builder.setContentTitle("알림 제목");
        builder.setContentText("알람 세부 텍스트");

        builder.setColor(Color.RED);
        // 사용자가 탭을 클릭하면 자동 제거
        builder.setAutoCancel(true);

        Intent _intent = new Intent(this,MainActivity.class);
        _intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        builder.setContentIntent(
                PendingIntent.getActivity(this,0,
                        _intent,
                        PendingIntent.FLAG_ONE_SHOT)
        );

        // 알림 표시
        NotificationManager notificationManager = (NotificationManager) this.getSystemService(Context.NOTIFICATION_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager.createNotificationChannel(new NotificationChannel("default", "기본 채널", NotificationManager.IMPORTANCE_DEFAULT));
        }

        // id값은
        // 정의해야하는 각 알림의 고유한 int값
        notificationManager.notify(1, builder.build());
    }

    private boolean isMyServiceRunning(Class<?> serviceClass) {

        ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);

        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            Log.d(TAG,"compare service" + service.service.getClassName() );
            if (serviceClass.getName().equals(service.service.getClassName())) {
                Log.d(TAG,"find service" + service.service.getClassName() );
                return true;
            }
        }
        Log.d(TAG,"not found");
        return false;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.d(TAG,"start App..");

        ((Button)findViewById(R.id.button_token)).setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Log.d(TAG,"click");
                        findToken();
                    }
                }
        );

        ((Button)findViewById(R.id.button_test_noti)).setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        createNotification();
                    }
                }
        );
        ((Button)findViewById(R.id.button_find_service)).setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        isMyServiceRunning(MyFcmService.class);

                    }
                }
        );


    }
}

File: lagacy/KtProactice_Json/app/src/main/java/com/example/ktproactice_json/MainActivity.kt
package com.example.ktproactice_json

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import kotlinx.android.synthetic.main.activity_main.*
import org.json.JSONArray
import org.json.JSONObject

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //파싱
        button_parse.setOnClickListener {
            val _JsonStr = "{\"r\":\"ok\",\"d\":[[\"12882072\",{\"type\":\"egcs\",\"ip\":\"128.0.0.0\",\"port\":62390,\"version\":110,\"at\":1601393794043}]]}"

            val _root = JSONObject(_JsonStr)
            println("r : ${_root["r"]}")

            val _data = _root["d"] as JSONArray
            val _item = _data[0] as JSONArray

            println("id : ${_item[0]}")
            println("data : ${_item[1]}")

            println("type : ${(_item[1] as JSONObject)["type"]}")

        }

        //생성
        button_stringify.setOnClickListener {
            val _root = JSONObject()
            _root.put("name","kotlin")
            _root.put("age",3)

            //배열 추가하기
            val _subroot = JSONArray()
            _subroot.put("pi")
            _subroot.put(3.14)
            _subroot.put(360)

            _root.put("data",_subroot)

            val _outStr = _root.toString()



            println(_outStr)
        }
    }
}